monitoring: use persistent storage for Prometheus

NicolasT opened the pull request:
These changes configure the Prometheus stateful services (i.e.,
Prometheus and AlertManager) to use a persistent volume for storage
instead of ephemeral `emptyDir` volumes.

As a result, these services don't 'work' out of the box after initial
deployment. We add some example `Volume` templates using the
`sparseLoopDevice` storage type to create proper volumes on a node after
deployment, which should allow the services to start.

Furthermore, the Vagrant deployment and CI are extended to create the
required `Volume`s such that the Prometheus services *are* available
after bootstrapping the system.

Note: during an earlier implementation of this feature (#1523), we ran
into a dependency issue between Prometheus deployment and the MetalK8s
UI deployment, which requires the first to be available. Instead, given
this PR, we use a hard-coded `NodePort` port number to access
Prometheus, not requiring the service to be available during bootstrap,
and point the UI to the workload-plane IP of the bootstrap node. This
ought to be improved once we use an `Ingress` to access the UI.

Fixes: #1515
See: https://github.com/scality/metalk8s/issues/1515
See: https://github.com/scality/metalk8s/pull/1523

NicolasT committed:
monitoring: use persistent storage for Prometheus

These changes configure the Prometheus stateful services (i.e.,
Prometheus and AlertManager) to use a persistent volume for storage
instead of ephemeral `emptyDir` volumes.

As a result, these services don't 'work' out of the box after initial
deployment. We add some example `Volume` templates using the
`sparseLoopDevice` storage type to create proper volumes on a node after
deployment, which should allow the services to start.

The `metalk8s-prometheus` *StorageClass* that used to be provisioned as
part of the `volumes` add-on is now deployed as part of the
`prometheus-operator` add-on. The rationale is as follows: the `volumes`
add-on (i.e., the operator we deploy) doesn't require any particular
*StorageClass* to be present, and as such should not deploy any
*StorageClass*es with it. Instead, an application which consumes
*PersistentVolume*s of a particular *StorageClass* must deploy these
*StorageClass*es with it, in casu the `prometheus-operator` add-on.

Furthermore, the Vagrant deployment and CI are extended to create the
required `Volume`s such that the Prometheus services *are* available
after bootstrapping the system.

Note: during an earlier implementation of this feature (#1523), we ran
into a dependency issue between Prometheus deployment and the MetalK8s
UI deployment, which requires the first to be available. Instead, given
this PR, we use a hard-coded `NodePort` port number to access
Prometheus, not requiring the service to be available during bootstrap,
and point the UI to the workload-plane IP of the bootstrap node. This
ought to be improved once we use an `Ingress` to access the UI.

Fixes: #1515
See: https://github.com/scality/metalk8s/issues/1515
See: https://github.com/scality/metalk8s/pull/1523

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

NicolasT commented:
@bert-e wait

NicolasT commented:
Once #1711 is in, this could be picked up again.

slaperche-scality committed:
test/volume: use prefix for volumes names

Like for StorageClass, we shouldn't blindly delete everything, as we may
have some objects (volumes in this case) required by the system.
Thus, like for StorageClass, we use a common prefix (poor's man
namespacing) for all the volume created by our tests.

Refs: #1681
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>
(cherry picked from commit cdbf8cdf3c0404809bd3ed04d0a50c4e30498b7f)

NicolasT commented:
@bert-e wait

NicolasT commented:
TODOs:

- [x] Docs

slaperche-scality commented on eve/create-volumes.sh:
```diff
@@ -0,0 +1,72 @@
+echo 'Waiting for Prometheus to be running'
+for _ in $(seq 1 60); do
+    PHASE=$(kubectl -n metalk8s-monitoring get pod prometheus-prometheus-operator-prometheus-0 -o jsonpath="{.status.phase}")
+    if [ "x${PHASE}" = "xRunning" ]; then
```
why the `x`?

NicolasT replied:
Old habits don't go away easily... https://unix.stackexchange.com/a/136632

slaperche-scality replied:
Ha good to know, thanks for the pointer :slightly_smiling_face:

slaperche-scality commented on eve/create-volumes.sh:
```diff
@@ -0,0 +1,72 @@
+    fi
+    sleep 1
+done
+[ $OK -eq 1 ] || (echo "Prometheus not Running"; false)
```
I guess the `wait for PV` and `wait for Pod` could be in function, and then we would call each of them twice, but meh it's shell scripts anyway…

NicolasT replied:
Not worth the trouble IMHO.

slaperche-scality replied:
Yeah, especially if we are going to rewrite this in Python at some point :stuck_out_tongue:

slaperche-scality commented on salt/metalk8s/addons/prometheus-operator/deployed/storageclass.sls:
```diff
@@ -1,14 +1,18 @@
-#!jinja | kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
+#! kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```
the move (and update) could have been in a dedicated commit, to explain the rationale behind the change/move.

Could also edit the current commit message to explain it.

NicolasT replied:
I'll amend the commit message. Not fond of doing moves like this (which don't bring any value per-se outside of the context of the actual commit doing the PV stuff and such) in separate commits.

NicolasT committed:
Merge remote-tracking branch 'origin/development/2.4' into improvement/1515-prometheus-pvc

* origin/development/2.4:
  packages/debian: fix buildrepo
  Get a working endpoint to check etcd health
  packages/debian: make `distributions` a template file
  build/targets: export Renderer
  build/iso: generate product.txt within a FileTree
  Deactivate temporarily Cypress test
  Sync pillar after bootstrap upgrade downgrade
  build/targets: add support for env file to SerializedData
  build/targets: allow to specify uptodate for Target
  build/targets: fix docstring of Target
  build/iso: reduce code duplication by using FileTree
  build/targets: add support for source prefix in FileTree
  build/targets: support copy to the root for FileTree
  fixed the select display bug

Conflicts:
	buildchain/buildchain/iso.py

NicolasT commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye nicolast.
